segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * An in-memory implementation of cache for read index that maps chunk start offset to chunk name for recently used segments.
+ * Eviction is performed only when number of segments or chunks exceeds certain given limits.
+ * Each time items are evicted the generation is incremented.
+ * For each entry in cache, we keep track of the generation in which it was last accessed.
+ * During eviction items with oldest generation are evicted first until enough objects are evicted.
+ * The least accessed segments are removed entirely before removing chunks from more recently used segments.
+ * This calculation is "best effort" and need not be accurate.
+ */
+class ReadIndexCache {
```

Comment:
Yes.

Reply:
I plan to work bit more on  ReadIndexCache  
I have created #4902 For tracking improvements in cache eviction logic for ReadIndexCache in ChunkManager layer. This will be done as separate PR after first set of major changes.